pravega/pravega
# Issue 6180: Recursive delete scope

**Change log description**  
Add a new deleteScopeRecursive() API on Controller that deletes all entities inside a scope and the scope itself.
This API should use the ControllerEventProcessor framework to asynchronously process the scope delete operation.

**Purpose of the change**  
Fixes #6180 , #6456, #6175

**What the code does**  
As per the description on the issue, 

- Created a new API `deleteScopeRecursive()` to delete all entities inside a scope and the scope itself too.
- Created a new table `_system/deletingScopes` which will hold the list of all scopes being deleted
- Applied checks for all create/update operations for RG/Stream/KVT to check if the scope is present in `_system/deletingScopes` table then no operation should be allowed
- Created a new event `DeleteScopeEvent.java`
- Created a new task `DeleteScopeTask.java` which implements `ScopeTask.java` and calls `deleteScopeRecursive()` of `PravegaTablesScope.java` which deletes all tables associated with the scope and then deletes the scope and removes entry from the _system/deletingScopes.
- Added method `deleteScopeRecursive(scopeName)` in `StreamManagerImpl.java` which internally calls Controller's `deleteScopeRecursive()` method.
- Update all existing test cases as a new event was added
- Add new test cases to test the new method
- Integration test Scenarios cover:-
> API without a valid scope 

> API with a valid scope

> Validation of completion by checking the existence in both tables i.e, `deleting_scopes_table` and `scopes_table`

> Revalidation by checking errors as expected for create of RG/Stream/KVTs under the same scope


**How to verify it**  
Issue deleteScopeRecursive() API and after all Streams in Scope are deleted, but before the operation completes, create a new Stream under the scope. The create/update/delete process of stream/RG/KVT should not be allowed.
